ql/src/java/org/apache/hadoop/hive/ql/plan/ConditionalResolverMergeFiles.java

```diff
@@ -419,17 +454,79 @@ private AverageSize getAverageSize(FileSystem inpFs, Path dirPath) {
    */
   private long getMergeSize(FileSystem inpFs, Path dirPath, long avgSize) {
     AverageSize averageSize = getAverageSize(inpFs, dirPath);
-    if (averageSize.getTotalSize() < 0) {
+    return getMergeSize(averageSize.totalSize, averageSize.numFiles, avgSize);
+  }
+
+  private List<FileStatus> getManifestFilePaths(HiveConf conf, Path dirPath) throws IOException {
+    FileSystem manifestFs = dirPath.getFileSystem(conf);
+    List<String> filesKept;
+    List<FileStatus> pathsKept = new ArrayList<>();
+    try (FSDataInputStream inStream = manifestFs.open(new Path(dirPath, Utilities.BLOB_MANIFEST_FILE))) {
+      String paths = IOUtils.toString(inStream, Charset.defaultCharset());
+      filesKept = new ArrayList(Arrays.asList(paths.split(System.lineSeparator())));
```

Comment:
why such nesting, new ArrayList when Arrays.asList already returns a list

Reply:
To make the list mutable. We need to perform a remove operation to remove the tbl directory.

Reply:
use guava, `Lists.newArrayList(paths.split(System.lineSeparator())`

Reply:
Done.